File: 2023/day06/day6.js
fs = require("fs")
function part1(lines) {
    let times = lines[0].split(": ")[1].split(" ").map(i => parseInt(i))
    let distances = lines[1].split(": ")[1].split(" ").map(i => parseInt(i))
    let product = 1

    for(let i = 0; i < times.length; i++) {
        product *= quadratic(times[i], distances[i])
    }
    return product
}

function part2(lines) {
    let time = parseInt(lines[0].split(": ")[1].replace(/ +/g,''))
    let distance = parseInt(lines[1].split(": ")[1].replace(/ +/g,''))
    return quadratic(time, distance)
}

function quadratic(time, distance) {
    let value = 0
    for(let i = 0; i < time + 1; i++) {
        if (-(Math.pow(i,2)) + time * i > distance)  {
            value += 1
        }
    }
    return value
}

function main() {
    let lines = fs.readFileSync('day6.txt', 'utf-8').replace(/ +/g,' ').split("\r\n");
    console.log(part1(lines))
    console.log(part2(lines))
}

main()

File: 2023/day06/day6.py
from functools import reduce
import re

def part1(file):
    file = re.sub(' +', ' ', file).split("\n")
    times = list(map(int, file[0].split(": ")[1].split(" ")))
    distances = list(map(int, file[1].split(": ")[1].split(" ")))
    summation = 1

    for index in range(len(times)):
        summation *= reduce(lambda sum, j: sum + (1 if j > distances[index] else 0), quadratic(times[index]), 0)
        print(summation)
    return summation

def part2(file):
    file = re.sub(' +', ' ', file).split("\n")
    time = int(re.sub(' +', '', file[0].split(": ")[1]))
    distance = int(re.sub(' +', '', file[1].split(": ")[1]))

    print(time)
    print(distance)
    print(quadratic(time))

    return reduce(lambda sum, j: sum + (1 if j > distance else 0), quadratic(time), 0)

def quadratic(time):
    values = []
    for i in range(time + 1):
        values.append( -(i**2) + time * i)
    return values


def main():
    file = open("day6.txt","r").read()
    print("Part 1: " + str(part1(file)))
    print("Part 2: " + str(part2(file)))

if __name__ == "__main__":
    main()